## 生成器 ##
在函数中如果出现了yield关键字，那么该函数就不再是普通函数，而是生成器函数。生成器表达式能做的事情列表解析基本都能处理，只不过在需要处理的序列比较大时，列表解析比较费内存。但是生成器函数可以生产一个无线的序列，这样列表根本没有办法进行处理

- 生成器的执行流程：

```py
def zrange(n):
    print("beginning of zrange")
    i = 0
    while i < n :
        print("before yield",i)
        yield i
        i += 1
        print("after yield",i)
    print("endding of zrange" )

Zrange = zrange(3)
print(Zrange)
print("****************")

Zrange.__next__()
print("----------------")

Zrange.__next__()
print("++++++++++++++++")

Zrange.__next__()
print("================")

Zrange.__next__()
print("################")

```
*打印输出*
```py
<generator object zrange at 0x004B78C8>
****************
beginning of zrange
before yield 0
----------------
after yield 1
before yield 1
++++++++++++++++
after yield 2
before yield 2
================
after yield 3
endding of zrange
Traceback (most recent call last):
  File "E:/PycharmProjects/untitled/study/day4/122.py", line 27, in <module>
    Zrange.__next__()
StopIteration

```
###### 通过运行结果可以看出：
- 当调用生成器函数的时候，函数只是返回了一个生成器对象，并没有执行；
- 当__next__()方法第一次被调用的时候，生成器函数才开始执行，执行到yield语句处停止， yield语句处的参数(yield value)就是__next__()方法的返回值
- 当继续调用__next__()方法的时候，函数将接着上一次停止的yield语句处继续执行，并到下一个yield处停止，如果后面没有yield就抛出Stoplteration异常


#### yeild与return
在一个生成器中，如果没有return，则默认执行到函数完毕时返回StopIteration；
```py
>>> def test1():
...     yield 1
>>> test=test1()
>>> next(test) #第一次调用next(g)时，会在执行完yield语句后挂起，所以此时程序并没有执行结束。
1
>>> next(test) #程序试图从yield语句的下一条语句开始执行，发现已经到了结尾，所以抛出StopIteration异常。
Traceback (most recent call last):
  File "<input>", line 1, in <module>
StopIteration

```
如果遇到return,如果在执行过程中 return，则直接抛出 StopIteration 终止迭代。
```py
>>> def test2():
...     yield 'a'
...     return
...     yield 'b'
>>> test = test2() #程序停留在执行完yield 'a'语句后的位置。
>>> next(test)
'a'
>>> next(test) #程序发现下一条语句是return，所以抛出StopIteration异常，这样yield 'b'语句永远也不会执行。
Traceback (most recent call last):
  File "<input>", line 1, in <module>
StopIteration

```
如果在return后返回一个值，那么这个值为StopIteration异常的说明，不是程序的返回值。
生成器没有办法使用return来返回值。
```py
>>> def test3():
...     yield 'hello'
...     return 'world'
>>> test=test3()
>>> next(test)
'hello'
>>> next(test)
Traceback (most recent call last):
  File "<input>", line 1, in <module>
StopIteration: world

```

**生成器表达式：**
当序列过长，而每次只需要获取一个元素时，应当考虑使用生成器表达式而不是列表解析，生成器表达式的语法和列表解析一样，只不过生成器表达式是被()括起来的，而不是[]，如下：



这地是代码



生成器表达式并不是创建一个列表，而是返回一个生成器，这个生成器在每次计算出一个条目后，把这个条目“产生”(yield)出来，生成器表达式使用了“惰性计算”，只有在检索时才被赋值，所以在列表比较长的情况下使用内存上更有效
```py
test = ( i for i in range(10) if i%2 )

print("__iter__" in dir(test))
print("__next__" in dir(test))
print(sum(test))

【说明】生成器表达式产生的生成器，它自身是一个可迭代对象，同时也是迭代器本身

```

**递归生成器：**




**生成器的send()和close()方法**

- send(value):  

从前面了解到，__next__方法可以恢复生成器状态并继续执行，其实send()是除__next__外另一个恢复生成器的方法。

yield语句也可以是yield表达式，也就是说yield可以有一个值，而这个值就是send()方法的参数，所以send(None)和next()是等效的。同样，next()和send()的返回值都是yield语句处的参数（yielded value）

关于send()方法需要注意的是：调用send传入非None值前，生成器必须处于挂起状态，否则将抛出异常。也就是说，第一次调用时，要使用next()语句或send(None)，因为没有yield语句来接收这个值。
```py
def gen():
    value=0
    while True:
        receive=yield value
        if receive=='e':
            break
        value = 'got: %s' % receive

g=gen()
print(g.send(None))
print(g.send('aaa'))
print(g.send(3))
print(g.send('e'))

```
*打印输出*
```py
0
got: aaa
got: 3
Traceback (most recent call last):
  File "E:/PycharmProjects/untitled/study/day4/122.py", line 72, in <module>
    print(g.send('e'))
StopIteration

```
###### 通过运行结果可以看出：

- 通过g.send(None)或者next(g)可以启动生成器函数，并执行到第一个yield语句结束的位置。此时，执行完了yield语句，但是没有给receive赋值。yield value会输出初始值0                  【注意】：在启动生成器函数时只能send(None),如果试图输入其它的值都会得到错误提示信息。
- 通过g.send(‘aaa’)，会传入aaa，并赋值给receive，然后计算出value的值，并回到while头部，执行yield value语句有停止。此时yield value会输出”got: aaa”，然后挂起。
- 通过g.send(3)，会重复第2步，最后输出结果为”got: 3″
- 当我们g.send(‘e’)时，程序会执行break然后推出循环，最后整个函数执行完毕，所以会得到StopIteration异常。


close():
这个方法用于关闭生成器，对关闭的生成器后再次调用next或send将抛出StopIteration异常。
```py
>>> def test4():
...     yield 1
...     yield 2
...     yield 3
... 
>>> test=test4()
>>> next(test)
1
>>> test.close()
>>> next(test)
Traceback (most recent call last):
  File "<input>", line 1, in <module>
StopIteration

```
throw()：
用来向生成器函数送入一个异常，可以结束系统定义的异常，或者自定义的异常。
throw()后直接抛出异常并结束程序，或者消耗掉一个yield，或者在没有下一个yield的时候直接进行到程序的结尾
```py
def gen():
    while True: 
        try:
            yield 'normal value'
            yield 'normal value 2'
            print('here')
        except ValueError:
            print('we got ValueError here')
        except TypeError:
            break
 
g=gen()
print(next(g))
print(g.throw(ValueError))  #根据上面的定义的except ValueError
print(next(g))
print(g.throw(TypeError))  #根据上面的定义的except TypeError

```
*打印输出*
```py

normal value
we got ValueError here
normal value
normal value 2
Traceback (most recent call last):
  File "E:/PycharmProjects/untitled/study/day4/122.py", line 89, in <module>
    print(g.throw(TypeError))
StopIteration
```
###### 通过运行结果可以看出：
- print(next(g))：会输出normal value，并停留在yield ‘normal value 2’之前。
- 由于执行了g.throw(ValueError)，所以会跳过所有后续的try语句，也就是说yield ‘normal value 2’不会被执行，然后进入到except语句，打印出we got ValueError here。然后再次进入到while语句部分，消耗一个yield，所以会输出normal value。
- print(next(g))，会执行yield ‘normal value 2’语句，并停留在执行完该语句后的位置。
- g.throw(TypeError)：会跳出try语句，从而print(‘here’)不会被执行，然后执行break语句，跳出while循环，然后到达程序结尾，所以跑出StopIteration异常。



#### 总结

- 通过实现迭代器协议对应的__iter__()和next()方法，可以自定义迭代器类型。对于可迭代对象，for语句可以通过iter()方法获取迭代器，并且通过next()方法获得容器的下一个元素。
- 像列表这种序列类型的对象，可迭代对象和迭代器对象是相互独立存在的，在迭代的过程中各个迭代器相互独立；但是，有的可迭代对象本身又是迭代器对象，那么迭代器就没法独立使用。
- itertools模块提供了一系列迭代器，能够帮助用户轻松地使用排列、组合、笛卡尔积或其他组合结构。
- 生成器是一种特殊的迭代器，内部支持了生成器协议，不需要明确定义__iter__()和next()方法。
- 生成器通过生成器函数产生，生成器函数可以通过常规的def语句来定义，但是不用return返回，而是用yield一次返回一个结果。

## 迭代器 ##
- 可以被__next__()函数调用并不断返回下一个值的对象称为迭代器；
- 迭代器是一个带状态的对象，他能在你调用next()方法的时候返回容器中的下一个值，任何实现了__iter__和__next__()（python2中实现next()）方法的对象都是迭代器，__iter__返回迭代器自身，__next__返回容器中的下一个值，如果容器中没有更多元素了，则抛出StopIteration异常，至于它们到底是如何实现的这并不重要。所以，迭代器就是实现了工厂模式的对象，它在你每次你询问要下一个值的时候给你返回

```py
>>> a = [1,2,3]
>>> dir(a)
['__add__', '__class__', '__contains__', '__delattr__',  '__delitem__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__gt__', '__hash__', '__iadd__', '__imul__', '__init__', '__iter__', '__le__', '__len__', '__lt__', '__mul__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__reversed__', '__rmul__', '__setattr__', '__setitem__', '__sizeof__', '__str__', '__subclasshook__', 'append', 'clear', 'copy', 'count', 'extend', 'index', 'insert', 'pop', 'remove', 'reverse', 'sort']
【说明】a是可迭代对象，但不是迭代器，因为dir查看支持的方法没有__next__方法

```


#### 总结

- 容器是一系列元素的集合，str、list、set、dict、file、sockets对象都可以看作是容器，容器都可以被迭代（用在for，while等语句中），因此他们被称为可迭代对象,但却不是迭代器，想把list、dict、str等可迭代对象变为迭代器可以使用iter()方法； 
```py
>>> a = [1,2,3]
>>> iter(a)
<list_iterator object at 0x029B89F0>
>>> b = iter(a)
>>> b.__next__()
1
>>> b.__next__()
2
>>> b.__next__()
3

```
Python的for循环本质上就是通过不断调用__next__函数实现的，例如
```py
for x in [1,2,3,4,5]:
    pass

就相当于：
#首先获得Iterator对象：
it = iter([1,2,3,4,5])

#for循环：
while True:
    try:
        #获得下一个值：
        x = next(it)
    except StopIteration:
        #遇到StopIteration就退出循环
        break

```

- 可迭代对象实现了__iter__方法，该方法返回一个迭代器对象；
- 迭代器持有一个内部状态的字段，用于记录下次迭代返回值，它实现了__next__和__iter__方法，迭代器不会一次性把所有元素加载到内存，而是需要的时候才生成返回结果；
- 生成器是一种特殊的迭代器，它的返回值不是通过return而是用yield；

```py
>>> from collections import Iterator
>>> isinstance(( x for x in range(5)), Iterator)
True

```
包括读取文件readlines也同样使用了迭代器：
```py
for i in f
其中f中会有next方法
```
- 迭代器代码程序是已经实现完成的，无需编程者自行编写代码，只需要了解迭代器概念即可


### 迭代器和生成器总结

- 生成器就是一种迭代器，可以使用for进行迭代。
- 第一次执行next(generator)时，会执行完yield语句后程序进行挂起，所有的参数和状态会进行保存。再一次执行next(generator)时，会从挂起的状态开始往后执行。在遇到程序的结尾或者遇到StopIteration时，循环结束。
- 可以通过generator.send(arg)来传入参数，这是协程模型。
- 可以通过generator.throw(exception)来传入一个异常。throw语句会消耗掉一个yield。可以通过generator.close()来手动关闭生成器。
- next()等价于send(None)